### DISCUSSION TITLE
[2024-04-15] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by farmaazon** [enso-org/enso#9698] [Weekly Demo Day]
My reduced iteration was entirely spent on #9691; Reproduced the closing websocket during hibernation; the reconnecting part was easy, but I wanted also at least initialize the protocol again, and that required a significant refactoring.

The PR in review make reconnections and re-initialization, but no more state update, so IDE still does not work properly after hibernation - the hibernation saga ends with #8522 and #8523 (the latter likely already working).

Also, a tiny fix: #9676 to not include `GUI`

**Comment by vitvakatu** [comment]
- https://github.com/enso-org/enso/pull/9612 – small fix for color picker.
- https://github.com/enso-org/enso/pull/9615 – runaway CSS rule was breaking documentation panel.
- https://github.com/enso-org/enso/pull/9616 – small fix for dropdown paddings, Pawel later fixed the rest.
- https://github.com/enso-org/enso/pull/9678 – we now display parenthesis for Group AST (if it is not top-level). 
- https://github.com/enso-org/enso/pull/9696 – small issue with node placement.
- https://github.com/enso-org/enso/pull/9677 – a quite complicated case of (mis-)interperting stale method call info in widget tree, causing some argument placeholders to be duplicated. Thanks to the discussion with the whole GUI team, we came to a reasonable solution.
- https://github.com/enso-org/enso/issues/9690 – *extremely* annoying dev server issue that happens on my machine only. Caused by some dependency update a month ago, still investigating.

**Comment by kazcw** [comment]
- #9633--Identify errors likely propagated from input nodes; hide them by default, like warnings:

https://github.com/enso-org/enso/assets/1047859/4b1b5e3d-c236-40d7-a3e7-e6ab8182ecd5

- #9671--Conditionalize a major parser dependency, and skip it in JNI/WASM builds (it is only needed in tests and CLI debugging utils).

**Comment by mwu-tow** [comment]
Landed:
* `clap` fixed the regression I've reported, allowing us to bump it: https://github.com/enso-org/enso/pull/9595 
* Nightly checks workflow landed: https://github.com/enso-org/enso/pull/9580
* JSON Assets description is part of the release: https://github.com/enso-org/enso/pull/9564
* Updated FlatBuffers dependency: https://github.com/enso-org/enso/pull/9654
* Updates for the Rust toolchain in the [`ci` repository](https://github.com/enso-org/ci)

In progress:
* Replacing NSIS installer — will allow bumping the electron-builder.

**Comment by Frizi** [comment]
- https://github.com/enso-org/enso/pull/9657 - Improved hot-reloading when working on widget definitions

https://github.com/enso-org/enso/assets/919491/8e6d5a67-68ec-4353-8235-32657b32e2ec

- https://github.com/enso-org/enso/pull/9685 - fixed issues caused by animating nodes on selection. Now animations are smooth.

https://github.com/enso-org/enso/assets/919491/e85fd68c-b2e8-4d58-90e1-4fd7b33f1c9b

- https://github.com/enso-org/enso/issues/9315 - Proper handling of node usages when a node is deleted

https://github.com/enso-org/enso/assets/919491/8551c901-4bef-4a7a-bfb6-10578f0c125a

- https://github.com/enso-org/enso/pull/9651 - Implemented local storage preservation of graph viewports on page reloads, separate for each project and function

https://github.com/enso-org/enso/assets/919491/fedce111-ea79-463f-a543-da3ecce28bf5

**Comment by JaroslavTulach** [comment]
- `25%` static compiler speed up on `longMethodWithLotOfLocalVars` - #9620
- Better AI integration - #9679
- Suggestion database reliably sent to the IDE  -  #9634
- JSON via Jackson: #9591
- Work on Y.js server in JVM is _almost finished_: https://github.com/enso-org/enso/pull/9528
- No known _"Autoscoped constructors problem"_ in the engine: https://github.com/enso-org/enso/issues/9614#issuecomment-2039135782
- `Table.join` PoC works - see **below** - #9410

Imagine you have two tables. One with columns `["a", "b", "e"]` and second with columns `["c", "b", "e"]`...

![two tables joined](https://github.com/enso-org/enso/assets/26887752/b1285b07-6ff1-4370-9913-30ece7d681c5)

After adding an `on` condition, the widget offers only `"b"` and `"e"`...

![on widget can access self and right](https://github.com/enso-org/enso/assets/26887752/1d5ebbd3-ea1b-46a4-9a97-e1e31c223470)

Why? Because the [Table.Widgets_Helper can access](https://github.com/enso-org/enso/pull/9410/files#r1525067756) values of arguments of the `join` function. If zeroth argument or argument named `right` is a `Table`, then the widgets can compare columns of `self` and this table and **offer only meaningful suggestions** - e.g. only columns that are present in both tables.

This is status as of https://github.com/enso-org/enso/pull/9410/commits/81aa2160c99a8a0ba3484f167a3c917e0190981a and it seems to be stable - e.g. you should be able to see the same on your computer after running `sbt:enso> runProjectManagerDistribution`
